lib/http2/connection.c

```diff
@@ -156,6 +149,17 @@ static void on_idle_timeout(h2o_timer_t *entry)
 
 static void update_idle_timeout(h2o_http2_conn_t *conn)
 {
+    if (!h2o_timer_is_linked(&conn->_graceful_shutdown_timeout)) {
+        h2o_linklist_unlink(&conn->super._conns);
+        if (0 == conn->num_streams.priority.open + conn->num_streams.priority.half_closed + conn->num_streams.pull.open +
+                conn->num_streams.pull.half_closed + conn->num_streams.push.open + conn->num_streams.push.half_closed) {
+            // all streams are idle
+            h2o_linklist_insert(&conn->super.ctx->_conns.idle, &conn->super._conns);
+        } else {
+            h2o_linklist_insert(&conn->super.ctx->_conns.active, &conn->super._conns);
+        }
+    }
```

Comment:
I think it might be possible to move this logic into `h2o_http2_stream_set_state`. The function is responsible for maintaining stream counts.